File: app/src/main/java/org/tasks/tags/TagPickerViewModel.java
package org.tasks.tags;

import static com.google.common.collect.Iterables.any;
import static org.tasks.Strings.isNullOrEmpty;

import androidx.annotation.Nullable;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import javax.inject.Inject;
import org.tasks.data.TagData;
import org.tasks.data.TagDataDao;
import org.tasks.tags.CheckBoxTriStates.State;

@SuppressWarnings({"WeakerAccess", "RedundantSuppression"})
public class TagPickerViewModel extends ViewModel {

  private final MutableLiveData<List<TagData>> tags = new MutableLiveData<>();
  private final CompositeDisposable disposables = new CompositeDisposable();

  @Inject TagDataDao tagDataDao;

  private final Set<TagData> selected = new HashSet<>();
  private final Set<TagData> partiallySelected = new HashSet<>();
  private String text;

  public void observe(LifecycleOwner owner, Observer<List<TagData>> observer) {
    tags.observe(owner, observer);
  }

  public void setSelected(List<TagData> selected, @Nullable List<TagData> partiallySelected) {
    this.selected.addAll(selected);
    if (partiallySelected != null) {
      this.partiallySelected.addAll(partiallySelected);
    }
  }

  public ArrayList<TagData> getSelected() {
    return new ArrayList<>(selected);
  }

  ArrayList<TagData> getPartiallySelected() {
    return new ArrayList<>(partiallySelected);
  }

  public String getText() {
    return text;
  }

  public void search(String text) {
    if (!text.equalsIgnoreCase(this.text)) {
      disposables.add(
          Single.fromCallable(() -> tagDataDao.searchTags(text))
              .subscribeOn(Schedulers.io())
              .observeOn(AndroidSchedulers.mainThread())
              .subscribe(this::onUpdate));
    }
    this.text = text;
  }

  private void onUpdate(List<TagData> results) {
    if (!isNullOrEmpty(text) && !any(results, t -> text.equalsIgnoreCase(t.getName()))) {
      results.add(0, new TagData(text));
    }
    tags.setValue(results);
  }

  @Override
  protected void onCleared() {
    super.onCleared();
    disposables.dispose();
  }

  State getState(TagData tagData) {
    if (partiallySelected.contains(tagData)) {
      return State.PARTIALLY_CHECKED;
    }
    return selected.contains(tagData) ? State.CHECKED : State.UNCHECKED;
  }

  State toggle(TagData tagData, boolean checked) {
    if (tagData.getId() == null) {
      tagData = new TagData(tagData.getName());
      tagDataDao.createNew(tagData);
    }

    partiallySelected.remove(tagData);

    if (checked) {
      selected.add(tagData);
      return State.CHECKED;
    } else {
      selected.remove(tagData);
      return State.UNCHECKED;
    }
  }
}

File: app/src/main/java/org/tasks/tags/TagPickerViewModel.kt
package org.tasks.tags

import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.state.ToggleableState
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import org.tasks.data.TagData
import org.tasks.data.TagDataDao
import javax.inject.Inject

@HiltViewModel
class TagPickerViewModel @Inject constructor(
        private val tagDataDao: TagDataDao
) : ViewModel() {

    private val tags = MutableLiveData<List<TagData>>()
    private val selected: MutableSet<TagData> = HashSet()
    private val partiallySelected: MutableSet<TagData> = HashSet()

    val searchText: State<String>
        get() = _searchText
    private val _searchText = mutableStateOf("")

    val tagToCreate: State<String>
        get() = _tagToCreate
    private val _tagToCreate = mutableStateOf("")

    fun observe(owner: LifecycleOwner, observer: (List<TagData>) -> Unit) =
            tags.observe(owner, observer)

    /* The property to access selected tags list from the @Composable activity */
    val tagsList: MutableLiveData<List<TagData>>
        get() = tags

    fun setSelected(selected: List<TagData>, partiallySelected: List<TagData>?) {
        this.selected.addAll(selected)
        if (partiallySelected != null) {
            this.partiallySelected.addAll(partiallySelected)
        }
    }

    fun getSelected() = ArrayList(selected)

    fun getPartiallySelected() = ArrayList(partiallySelected)

    fun search(newText: String) {
        if (newText == "" || !newText.equals(_searchText.value, ignoreCase = true)) {
            viewModelScope.launch {
                val results = tagDataDao.searchTags(newText)
                onUpdate(newText, results.toMutableList())
            }
        }
        _searchText.value = newText
    }

    private fun onUpdate(newText: String, results: MutableList<TagData>) {
        val sorted = results
            .sortedWith { l, r ->
                val lSelected = selected.contains(l) || partiallySelected.contains(r)
                val rSelected = selected.contains(r) || partiallySelected.contains(r)
                if (lSelected && !rSelected) {
                    -1
                } else if (rSelected) {
                    1
                } else {
                    0
                }
            }
            .toMutableList()
        if ( newText != "" && !results.any {newText.equals(it.name, ignoreCase = true) } )
            _tagToCreate.value = newText
        else
            _tagToCreate.value = ""
        tags.value = sorted
    }

    fun getState(tagData: TagData): ToggleableState {
        if (partiallySelected.contains(tagData)) {
            return ToggleableState.Indeterminate
        }
        return if (selected.contains(tagData)) ToggleableState.On else ToggleableState.Off
    }

    suspend fun toggle(tagData: TagData, checked: Boolean): ToggleableState {
        var tagData = tagData
        if (tagData.id == null) {
            tagData = TagData(tagData.name)
            tagDataDao.createNew(tagData)
        }
        partiallySelected.remove(tagData)
        return if (checked) {
            selected.add(tagData)
            ToggleableState.On
        } else {
            selected.remove(tagData)
            ToggleableState.Off
        }
    }

    suspend fun createNew(name: String) {
        val tagData = TagData(name)
        tagDataDao.createNew(tagData)
        selected.add(tagData)
        search("")
    }
}